### ISSUE TITLE
ONNX release 1.7

### ISSUE THREAD

**Issue by linkerzhang** [onnx/onnx#2614]
Per ONNX community agreement, ONNX 1.7 release is being prepared. Key dates as below.
1.	PRs (for 1.7) merging due date – 2/25/2020.
2.	Release branch (for 1.7) creation date – 2/26/2020.
3.	1.7 release date – 3/2/2020.

Please active PR owners, mark your PR with “1.7 release” tag (in PR subject) and actively pushing the merge before the due date.

**Comment by linkerzhang** [comment]
Hi ONNX partners, 
Due to some key PRs not merged yet, per the community agreement, ONNX 1.7 release will be pushed out for one week. Key dates updated as below,
a.	PRs (for 1.7) merging due date – 3/3/2020.
b.	Release branch (for 1.7) creation date – 3/4/2020.
c.	1.7 release date – 3/9/2020.

**Comment by linkerzhang** [comment]
@chinhuang007 how do you think of tracking the issues left for 1.7 release here please? (it's a little bit hard to track issues in the gitter message room). Thank you!

**Comment by codemzs** [comment]
### <div align="center">Issues that were fixed to ensure ONNX 1.7 ships with quality.</div>

| **Operator**  | **Offending PR**  | **Issue** | **Fix Description** | **Fix PR** |
|:-:|:-:|:-:|:-:|:-:|
| SoftmaxCrossEntropyLoss | https://github.com/onnx/onnx/pull/2573  | Incorrect label tensor type as int32/int64 in the spec instead of float. | Update the spec. | https://github.com/onnx/onnx/pull/2667 |
| SoftmaxCrossEntropyLoss | https://github.com/onnx/onnx/pull/2573  | Spec is missing `ignore_index` attribute, effectively rendering it useless for use with PyTorch models. Real world use case with HuugingFace model uses `ignore_index` equal to -1. | Update the spec, reference implementation, function body, tests.  | https://github.com/onnx/onnx/pull/2680 , https://github.com/onnx/onnx/pull/2703 , https://github.com/onnx/onnx/pull/2700, https://github.com/onnx/onnx/pull/2696 |
| SoftmaxCrossEntropyLoss | https://github.com/onnx/onnx/pull/2573  | Incorrect reference implementation.  | Fix bugs in reference implementation and handle 3-D, 4-D, and other cases.  | https://github.com/onnx/onnx/pull/2690 |
| SoftmaxCrossEntropyLoss | https://github.com/onnx/onnx/pull/2573  | Function Body definition produces inaccurate results. | <ul><li>Max operator was being used to find max class score per example instead of ReduceMax(axis = 1). Max is used for finding element wise max across multiple tensors, not sure why this was considered for a single tensor and to find max for every class in each example.</li></ul>  <ul><li>To sum up scores per class ReduceSum was being used without specifying axis = 1, this was summing across all axis instead of just along the classes.</li></ul> | https://github.com/onnx/onnx/pull/2725 |
| SoftmaxCrossEntropyLoss | https://github.com/onnx/onnx/pull/2573  | Function body does not handle optional output _log_probability_ | Handle optional output _log_probability_ | https://github.com/onnx/onnx/pull/2725 |
| SoftmaxCrossEntropyLoss | https://github.com/onnx/onnx/pull/2573  | Reference implementation does not handle optional output _log_probability_ | Handle optional output _log_probability_ | https://github.com/onnx/onnx/pull/2725 |
| SoftmaxCrossEntropyLoss | https://github.com/onnx/onnx/pull/2573  | No tests for optional output _log_probability_ | Add ~60 tests for optional output _log_probability_ | https://github.com/onnx/onnx/pull/2725 |
| NegativeLogLikelihoodLoss | https://github.com/onnx/onnx/pull/2551  | Function body does not handle target values of types int64  | Fix function body to handle target values of all possible supported types, previously it was only handling int32 and not int64, this was caught when I ran the updated spec with ORT kernel tests.  | https://github.com/onnx/onnx/pull/2725 |
| NegativeLogLikelihoodLoss | https://github.com/onnx/onnx/pull/2551  | Incorrect Shape inference asserts causing crash in ORT. | Fix ShapeInference with incorrect assert on the dimensions of weights input (caught via ORT kernel tests) | https://github.com/onnx/onnx/pull/2573 |
| Celu | https://github.com/onnx/onnx/pull/2573  | `alpha` attribute passed as `float_value` instead of `tensor` type  | Fix function body to prevent failures in ORT backend. Attribute type was being passed as float_value but in the ORT backend we only accept attribute type of tensor. | https://github.com/onnx/onnx/pull/2573 |
| Dropout | https://github.com/onnx/onnx/pull/2568  | Default probability does not match that of PyTorch, TensorFlow and ORT  | Add `training_mode` input to Dropout spec to support training and inference mode by keeping default ratio as 0.5. | https://github.com/onnx/onnx/pull/2725 |
| Dropout | https://github.com/onnx/onnx/pull/2568  | Default probability does not match that of PyTorch, TensorFlow and ORT  | Add `training_mode` input to Dropout spec to support training and inference mode by keeping default ratio as 0.5. | https://github.com/onnx/onnx/pull/2725 |
| Dropout | https://github.com/onnx/onnx/pull/2568 | Incorrect reference implementation that did not have scaling factor and no option to return optional parameter `mask`  | Rewrite reference implementation by adding scaling factor and provision to return mask. | https://github.com/onnx/onnx/pull/2725 |
| Dropout | https://github.com/onnx/onnx/pull/2568 | Inadequate test coverage by only having two tests.  | Rewrite tests and increase the count to 10 to cover all possible scenarios. Update ORT backend to have a special provision(count zeros instead of absolute comparison) to handle dropout tests as the mask generated will be different from reference implementation. | https://github.com/onnx/onnx/pull/2725 |
| UnfoldToDepth | https://github.com/onnx/onnx/pull/2616 , https://github.com/onnx/onnx/pull/2722 | Inadequate validation of this operator spec.  | Revert the spec. | https://github.com/onnx/onnx/pull/2749 |
| Inverse | https://github.com/onnx/onnx/pull/2578 , https://github.com/onnx/onnx/pull/2706 | Inadequate validation of this operator spec.  | Revert the spec. | https://github.com/onnx/onnx/pull/2747 |
| MeanSquaredDistance | https://github.com/onnx/onnx/pull/2570 | Inadequate validation of this operator spec.  | Revert the spec. | https://github.com/onnx/onnx/pull/2748 |
| BatchNormalization(12) | https://github.com/onnx/onnx/pull/2568 , https://github.com/onnx/onnx/pull/2674 | Inadequate validation of this operator spec.  | Revert the spec. | https://github.com/onnx/onnx/pull/2750 |

CC: @linkerzhang , @chinhuang007 , @prasanthpul , @SherlockNoMad , @ebarsoum , @faxu

**Comment by ebarsoum** [comment]
For 1.7 release, we discussed the below two options:

1. Rename the ‘training’ domain to ‘preview’, and in the release note explicitly says that training is in preview mode and it isn’t part of the spec. Same as we said in the last workshop.
2. Add cmake flag to not build operators under the training domain and manually copy the old generated MD documentation into a separate MD file that will be checked-in with the release.

The problem in (1) is that we didn't yet finalize the `preview` proposal and it might take a month to do so. The advantage is that the change is minimum and align to what we announced in the workshop.

The problem in (2) is more works that is throw away, advantage operator under training domain wont be part of ONNX binary.

CC: @linkerzhang , @chinhuang007 , @prasanthpul , @SherlockNoMad , @smessmer , @ebarsoum , @faxu, @yuanbyu, @gramalingam , @wschin , @codemzs , @postrational

**Comment by chinhuang007** [comment]
@ebarsoum Thanks for the summary!

Given that 1.7 release has been delayed for 2 months due to various issues as @codemzs listed above, I think another couple of weeks might not be a big deal (of course I can't speak for onnx runtime which might be delayed as well). The important thing is have a high quality release with a clear scope.

Personally I don't like throw-away code. Even for 2, it might take some time because someone has to write code and go through another PR. So I am leaning toward 1 and wonder if we could expedite it with an agreed target date from community.

**Comment by gramalingam** [comment]
@chinhuang007 , that sounds good! For (1), what would you like to see formalized/finalized regarding the "preview" proposal? We can then try to make progress on it as fast as possible.

**Comment by wschin** [comment]
The use of preview domain can be more strict in the beginning and we gradually relax it if we want. For example, we can say, in 1.7,

1. this preview domain is only for containing training operators.
2. those operators in preview will be removed from preview domain and moved to training domain in the future.

In the future, we can probably add

3. operators which doesn't map to individual operators in existing frameworks should go to preview first.
4. an operator can only exist in preview for 3 months. If no further development happens, we remove it.

**Comment by chinhuang007** [comment]
@gramalingam The preview proposal is a brand new concept, at least in ONNX. Like anything new feature introduced into an open source community, I think we should clearly articulate

- why we have it, ie what problems it helps to solve
- how to use it, such as just specify an op as in onnx.preview domain?
- what is the impact to ONNX build and CI?
- what are the differences between a regular op vs a preview op in terms of quality criteria, unit tests, backend validation, versioning
- how to decide my op should be in preview or not
- how long can an op stay in preview
- what are the possible destinations for a preview op (regular op and removed?)
- how to move a preview op to a regular op
- are preview ops part of the standard backend test that all backend (converters and runtimes) should support?
- what ONNX partners should expect from preview ops
- any differences in PR review and approval between a preview op and a regular op

**Comment by gramalingam** [comment]
(a) It exists as a staging ground to propose an op, get feedback from the community, to iterate over the specification and finalize it.
(b) Yes, it will be in a special domain we choose.
(c) I prefer it have to no impact on ONNX build and CI.
(d) This is one main difference (quality criteria). Regular op has higher standards. The experimental op is expected to improve in quality, unit cases, backend validation over time until it meets the bar for regular op.
(e) I think the criteria should be quality as in (d) and utility/value as agreed by the committee/reviewers.
(f) I don't have an answer to how long an op can stay in preview.
(g) yes, a preview op can become regular op or be dropped
(h) As before: it needs to clear the bar for standard op.
(i) It is not required, I think,
(j) Not sure what this means
(k) Yes, as above. Higher bar for regular op

**Comment by ebarsoum** [comment]
Hi @chinhuang007 

Those are good questions, I don't have all the answer, because we need to agree on the detail in which the write up is in progress. 

But here my current thinking:

- why we have it, ie what problems it helps to solve
Give opportunity for runtime writer to implement the spec and provide feedbacks for operators that isn't part of any frameworks (such as training and others in the future)

- how to use it, such as just specify an op as in onnx.preview domain?
Yes, same as ONNXML.

- what is the impact to ONNX build and CI?
Same as core operators, they will have unittest and it will be part of CI.

- what are the differences between a regular op vs a preview op in terms of quality criteria, unit tests, backend validation, versioning
Quality should be pretty close, the main difference is that those operators aren't part of existing framework. So new signature that need some feedbacks and iteration from the community. The OP will have version for tracking, but preview Ops aren't part of the spec release, and can be removed. 

- how to decide my op should be in preview or not
Only operators that don't have equivalent in existing framework or the equivalent is different.

- how long can an op stay in preview
TBD, this something we can added later once approved. The time will be bounded.

- what are the possible destinations for a preview op (regular op and removed?)
Yes, regular or removed OP. Or a different OP based on feedbacks.

- how to move a preview op to a regular op
Change domain and reset the OPSet version.

- are preview ops part of the standard backend test that all backend (converters and runtimes) should support?
No, there is no ask for converter or runtime to support it. But it still need to have reference implementation and backend test as standard OP.

- what ONNX partners should expect from preview ops
Review it and provide feedbacks, and if they have runtime it will be good to try and see if something is missing.

- any differences in PR review and approval between a preview op and a regular op
Mostly not, the only difference is that those OPs might not have equivalent in existing framework. And you might not have model to support them for example if TFFoo is different than PTFoo which is different then mxnetFoo, then we might come up with GeneralFoo that supersede them.

Thanks,
Emad

**Comment by linkerzhang** [comment]
My two cents:

1. I agree that we'd invest more time/energy on figuring out the definition of "preview" if it's needed.
2. I agree to not block current 1.7 release following the 2nd option.

The 2nd option is not that expensive. The work includes:
1. add a cmake flag to disable the c++ registration of the ops which we think not ready for production quality (by default it's disabling).
No codes thrown away.
2. Enable the cmake flag to generate an operators.md and rename it as Operators_preview.md and check in this new Operator_preview.md only.
3. When doing release, disable the flag (by default) to build the binary and release it.

**Comment by wschin** [comment]
@linkerzhang, Gradient and GraphCall are the center of training -- Gradient for backward pass and GraphCall for forward pass. It looks strange if optimizers come without Gradient and GraphCall --- it can be interpreted as something like `we have optimizer but we can't compute gradient...`. I personally think optimizers should go wherever Gradient goes.

This preview thing happens everywhere. In GCC, in Pytorch, in Tensorflow, they just clearly indicate its preview status. Here is a message from GCC page:
```
Important: Because the ISO C++2a standard is still evolving, GCC's support is experimental.
No attempt will be made to maintain backward compatibility with implementations of C++
2a features that do not reflect the final standard.
```
I don't understand why ONNX can't have this concept.

**Comment by codemzs** [comment]
@wschin No one is disagreeing on having preview! We SHOULD! :) It is about how you want to execute it which needs more thought and not risk the release.

**Comment by wschin** [comment]
I think trying to change training spec is the root of this risk --- training components are tighten together. If you change one component, you need to consider its interaction with the entire rest of training spec. The minimal change is to move everything into a preview domain. Again, any change against training will need more time to think about and delay the release.

**Comment by chinhuang007** [comment]
@ebarsoum I am creating the release note and would like to a sentence or two to describe the preview domain. Can you please help? Thanks.

**Comment by codemzs** [comment]
Thanks all. We have released ONNX 1.7 on [PyPi](https://pypi.org/project/onnx/)

**Comment by chinhuang007** [comment]
ONNX 1.7 is released today, on PyPi. The v1.7.0 release and tag on github are also created. Thanks everyone for your help!! We can finally close this issue now.

**Comment by codemzs** [comment]
Thanks @chinhuang007 for your help with the release!!!

**Comment by gramalingam** [comment]
Thanks @chinhuang007 !

**Comment by linkerzhang** [comment]
Thank you @chinhuang007 so much for making the release! I'm closing this issue now.

**Comment by chinhuang007** [comment]
A Release working group is forming to help make future releases more efficiently, https://github.com/onnx/steering-committee/blob/master/meeting-notes/20200521.md. The kickoff meeting is scheduled as 10-10:30 am Wed. 6/3/3030, Pacific Time, https://zoom.us/j/7376656864. We will start with identifying high priority issues and action items in such as Windows build, Conda-forge, communication, verification, documentation, and planning/scoping. Please plan to join. Your contribution is highly needed and appreciated.

**Comment by chinhuang007** [comment]
@codemzs We are updating the release doc to reflect the process and steps for 1.7 release. One missing piece is the details of the verification work you have helped to accomplish. We would like to understand and document it so anyone with certain knowledge and skills can perform this critical step for release 1.8 and beyond.

I wonder if you can kindly summarize what you have done, in terms of:
- what is the target environment (Linux, OSX, Windows etc)
- what is the software stack (python, protobuf versions etc)
- how to run the verification tests

Thank you very much in advance!

I also added @askhade to this thread as she is a core contributor to the Release working group
